tests/mock_network/churn.rs

```diff
@@ -20,65 +20,105 @@ use routing::{
-        let (initial_size, max_drop) = sections[&pfx];
-        if drop_count.is_empty() {
-            break;
-        } else if drop_count.get(&pfx).is_none() {
+
+    // 10% probability that a node will be dropped.
+    let drop_probability = 0.1;
+
+    let mut sections = HashMap::<_, SectionInfo>::new();
+    for node in nodes.iter() {
+        let prefix = *node.our_prefix();
+        let info = sections.entry(prefix).or_default();
+        if node.inner.is_elder() {
+            info.initial_elder_count += 1;
+        } else {
+            info.initial_other_count += 1;
+        }
+    }
+
+    let mut dropped_indices = Vec::new();
+    let mut dropped_names = BTreeSet::new();
+
+    for (index, node) in nodes.iter().enumerate() {
+        if rng.gen_range(0.0, 1.0) >= drop_probability {
             continue;
         }
 
-        let early_terminate = max_per_pfx.map_or(false, |n| {
-            drop_count[&pfx] >= n || rng.gen_weighted_bool(drop_count.keys().len() as u32)
-        });
-        let normal_terminate =
-            ((drop_count[&pfx] + 1) * 3 >= initial_size) || (drop_count[&pfx] >= max_drop);
-        if early_terminate || normal_terminate {
-            let _ = drop_count.remove(&pfx);
+        let section = unwrap!(sections.get_mut(node.our_prefix()));
+        let elder_size = unwrap!(node.inner.elder_size());
+        let remaining = section.initial_elder_count + section.initial_other_count
+            - section.dropped_other_count
+            - section.dropped_elder_count;
+        let dropped = section.dropped_elder_count + section.dropped_other_count;
+
+        // Check the optional additional drop limit.
+        if let Some(max_per_pfx) = max_per_pfx {
+            if dropped + 1 > max_per_pfx {
+                continue;
+            }
+        }
+
+        // Don't drop any other node if an elder is already scheduled for drop.
+        if section.dropped_elder_count > 0 {
+            continue;
+        }
+
+        // Don't drop below elder_size nodes.
+        if remaining <= elder_size {
             continue;
         }
 
-        *unwrap!(drop_count.get_mut(&pfx)) += 1;
-        let dropped = nodes.remove(i);
-        assert!(dropped_nodes.insert(dropped.name()));
+        // If there already are other drops scheduled, make sure we remain with at least one
+        // more node above elder_size. This is because one of those other drops might trigger
+        // relocation of one of the existing elders and we wouldn't have anyone to replace it with
+        // otherwise.
+        if dropped > 0 && remaining <= elder_size + 1 {
+            continue;
+        }
+
+        if node.inner.is_elder() {
+            // Don't drop elder if a non-elder is already scheduled for drop.
+            if section.dropped_other_count > 0 {
+                continue;
+            }
+
+            section.dropped_elder_count += 1;
+        } else {
+            section.dropped_other_count += 1;
```

Comment:
just to confirm, so we can drop multiple non-elder members, right?

Reply:
Yes, that's the idea. So far it looks like it works, but let's wait for the soak test results.